### PR TITLE
Book security user

### PR DESCRIPTION
**Pull request by Nikname2303**

### CODE REVIEW

File: src/main/java/com/example/bookshop/annotation/FieldMatchesValidator.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import com.example.bookshop.dto.UserRegistrationRequestDto;
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+
+public class FieldMatchesValidator implements ConstraintValidator<FieldMatch, Object> {
+    private String[] fields;
+
+    @Override
+    public void initialize(FieldMatch constraintAnnotation) {
+        fields = constraintAnnotation.value();
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/annotation/FieldMatchesValidator.java]: fields is not used

File: src/main/java/com/example/bookshop/dto/UserRegistrationRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import com.example.bookshop.annotation.FieldMatch;
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.validator.constraints.Length;
+
+@Getter
+@Setter
+@FieldMatch({"password", "repeatPassword"})
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/dto/UserRegistrationRequestDto.java]: No reason to define values, you don`t use them. Please implement annotation to make it work with any 2 fields

File: src/main/java/com/example/bookshop/mapper/UserMapper.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import com.example.bookshop.dto.UserResponseDto;
+import com.example.bookshop.model.User;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
+public interface UserMapper {
+    UserResponseDto toResponseDto(User user);
+
+    User toModel(UserRequestDto requestDto);
+
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/mapper/UserMapper.java:15]:
```suggestion
```

File: src/main/java/com/example/bookshop/service/UserService.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package com.example.bookshop.service;
+
+import com.example.bookshop.dto.UserRegistrationRequestDto;
+import com.example.bookshop.dto.UserResponseDto;
+import com.example.bookshop.exception.RegistrationException;
+
+public interface UserService {
+    UserResponseDto register(UserRegistrationRequestDto requestDto) throws RegistrationException;
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/service/UserService.java]: RegistrationException is unchecked, it is defined in signature?

File: src/main/java/com/example/bookshop/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto) {
+        User user = new User();
+        user.setEmail(requestDto.getEmail());
+        user.setPassword(requestDto.getPassword());
+        user.setFirstName(requestDto.getFirstName());
+        user.setLastName(requestDto.getLastName());
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/service/impl/UserServiceImpl.java]: Is it possible to use userMapper?

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 6]
```diff
@@ -3,3 +3,5 @@ databaseChangeLog:
       file: db/changelog/changes/01-create-books-table.yaml
   - include:
       file: db/changelog/changes/02-add-column-isDeleted-to-books-table.yaml
+  - include:
+      file: db/changelog/changes/03-create-users-table.yaml
```
Comment (andrii-hoienko) [inline src/main/resources/db/changelog/db.changelog-master.yaml]:
```suggestion
      file: db/changelog/changes/03-create-users-table.yaml
      
```

File: src/main/java/com/example/bookshop/exception/CustomGlobalExceptionHandler.java [hunk 7]
```diff
@@ -34,6 +34,15 @@ public ResponseEntity<Object> handleException(Exception exception) {
                 .body(exception.getMessage());
     }
 
+    @ExceptionHandler({RegistrationException.class})
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/exception/CustomGlobalExceptionHandler.java]:
```suggestion
    @ExceptionHandler(RegistrationException.class)
```

File: src/main/java/com/example/bookshop/annotation/FieldMatch.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+@Documented
+public @interface FieldMatch {
+    String[] value();
+
+    String message() default "Passwords must be identical";
+
+    Class<?>[] groups() default {};
+
+    Class<? extends Payload>[] payload() default {};
+
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/annotation/FieldMatch.java:22]:
```suggestion
```

File: pom.xml [hunk 9]
```diff
@@ -76,6 +76,11 @@
 			<groupId>org.hibernate.validator</groupId>
 			<artifactId>hibernate-validator</artifactId>
 		</dependency>
+		<dependency>
+			<groupId>org.springdoc</groupId>
+			<artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
+			<version>2.1.0</version>
+		</dependency>
```
Comment (andrii-hoienko) [inline pom.xml]: This PR contains changes from 2 tasks, add User entity and Pagination, Sorting, Swagger. Please make PR only with relevant changes

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good job, but some improvements required

**Comment by Nikname2303** [comment]
I fixed problem with previous commits, but with merge, so I still have history from previous PR. But files changes have only relevant changes. On QnA we discussed about that

File: src/main/java/com/example/bookshop/annotation/FieldMatchesValidator.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    public void initialize(FieldMatch constraintAnnotation) {
+        ConstraintValidator.super.initialize(constraintAnnotation);
+    }
+
+    @Override
+    public boolean isValid(Object o, ConstraintValidatorContext constraintValidatorContext) {
+        UserRegistrationRequestDto userDto = (UserRegistrationRequestDto) o;
+        return userDto.getPassword().equals(userDto.getRepeatPassword());
+    }
+}
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/annotation/FieldMatchesValidator.java]: This still works only for 2 specific fields

File: src/main/java/com/example/bookshop/exception/RegistrationException.java [hunk 11]
```diff
@@ -0,0 +1,7 @@
+package com.example.bookshop.exception;
+
+public class RegistrationException extends Exception {
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/exception/RegistrationException.java]:
```suggestion
public class RegistrationException extends RuntimeException {
```

File: src/main/java/com/example/bookshop/service/UserService.java [hunk 12]
```diff
@@ -0,0 +1,9 @@
+package com.example.bookshop.service;
+
+import com.example.bookshop.dto.user.UserRegistrationRequestDto;
+import com.example.bookshop.dto.user.UserResponseDto;
+import com.example.bookshop.exception.RegistrationException;
+
+public interface UserService {
+    UserResponseDto register(UserRegistrationRequestDto requestDto) throws RegistrationException;
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/service/UserService.java]:
```suggestion
    UserResponseDto register(UserRegistrationRequestDto requestDto);
```

File: src/main/java/com/example/bookshop/service/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto) {
+        User savedUser = userRepository.save(userMapper.toModelFromRegisterDto(requestDto));
```
Comment (andrii-hoienko) [inline src/main/java/com/example/bookshop/service/impl/UserServiceImpl.java:24]: You should check is email valid

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Please read requirementsand comment cerafully

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good job!

**Review by boroda4436 — APPROVED** [review]

**Pull request merged**